File: jetpack_java/src/main/java/com/kunminx/jetpack_java/sample_01_lifecycles/ui/LifecycleLocationActivity.java


package com.kunminx.jetpack_java.sample_01_lifecycles.ui;

import android.content.Intent;
import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;

import com.kunminx.architecture.ui.BaseActivity;
import com.kunminx.architecture.ui.layout_manager.WrapContentLinearLayoutManager;
import com.kunminx.jetpack_java.R;
import com.kunminx.jetpack_java.common_data.Configs;
import com.kunminx.jetpack_java.common_ui.adapter.LocationAdapter;
import com.kunminx.jetpack_java.sample_01_lifecycles.domain.LifecycleLocationManager;

/**
 * TODO：本示例专注于提供 Lifecycle 组件的使用场景和示例，因而其他内容均保持 Android 引入 Jetpack 前的模样，
 * kotlin 模块提供同样的场景和基于 Kotlin 的写法，可对照查阅。
 * 并且本项目的示例从 sample_01 到 sample_05 是循序渐进地 Jetpack 化，
 * 如看完对 Jetpack 高频核心组件 "为什么要用"、"为什么要这样用" 有了一丝丝好奇心，可前往《Jetpack MVVM 精讲》和《Jetpack MVVM 最佳实践》项目查阅深度解析。
 * <p>
 * https://juejin.im/post/5dafc49b6fb9a04e17209922
 * <p>
 * https://github.com/KunMinX/Jetpack-MVVM-Best-Practice
 * <p>
 * <p>
 * Create by KunMinX at 19/10/16
 */

public class LifecycleLocationActivity extends BaseActivity {

    private RecyclerView mRecyclerView;
    private Toolbar mToolbar;
    private LocationAdapter mLocationAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_location_lifecycles);
        mToolbar = findViewById(R.id.toolbar);
        mToolbar.setNavigationOnClickListener(v -> finish());

        mRecyclerView = findViewById(R.id.rv);
        mRecyclerView.setLayoutManager(new WrapContentLinearLayoutManager(getApplicationContext()));

        mRecyclerView.setAdapter(mLocationAdapter = new LocationAdapter(getApplicationContext(), locationBean -> {
            Intent intent = new Intent();
            intent.putExtra(Configs.LOCATION_RESULT, locationBean.getLocationName());
            setResult(RESULT_OK, intent);
            finish();
        }));

        getLifecycle().addObserver(LifecycleLocationManager.getInstance());

        LifecycleLocationManager.getInstance().setILocationCallback(list -> {
            runOnUiThread(() -> {
                mLocationAdapter.submitList(list);
            });
        });
    }

}

File: jetpack_kotlin/src/main/java/com/example/jetpack_kotlin/sample_01_lifecycles/ui/LifecycleLocationActivity.kt
package com.example.jetpack_kotlin.sample_01_lifecycles.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.recyclerview.widget.RecyclerView
import com.example.jetpack_koltin.R
import com.example.jetpack_kotlin.common_data.Configs
import com.example.jetpack_kotlin.common_data.bean.LocationBean
import com.example.jetpack_kotlin.common_ui.adapter.LocationAdapter
import com.example.jetpack_kotlin.sample_01_lifecycles.domain.LifecycleLocationManager
import com.kunminx.architecture.ui.BaseActivity

/**
 * @author Flywith24
 * @date   2020/5/30
 * time   11:20
 * description
 */
class LifecycleLocationActivity : BaseActivity(R.layout.activity_lifecycles_location), LocationAdapter.OnItemClickListener {
    /**
     * 不推荐使用 Kotlin Synthetics
     * 可以使用 ViewBinding 和 功能更强大的 DataBinding 来替换 findViewById
     *
     * 本示例尽量只演示单个组件的使用，因此没有使用 ViewBinding 或 DataBinding
     *
     * 详情参考 https://juejin.im/post/5e8ef0bc518825736b749705#heading-17
     */
    private lateinit var mRecyclerView: RecyclerView
    private val mAdapter by lazy { LocationAdapter(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mRecyclerView = findViewById(R.id.rv)
        mRecyclerView.adapter = mAdapter

        lifecycle.addObserver(LifecycleLocationManager.newInstance())

        LifecycleLocationManager.newInstance().setILocationCallback { list ->
            runOnUiThread {
                mAdapter.submitList(list)
            }
        }
    }

    override fun onItemClick(locationBean: LocationBean) {
        setResult(Activity.RESULT_OK, Intent().putExtra(Configs.LOCATION_RESULT, locationBean.locationName))
        finish()
    }
}